File: secondo_anno/Algoritmi/Modulo_due/programmazione_dinamica/lezione_1/cammino.c
#include <stdio.h>
#include <stdlib.h>

int max(int a, int b);
int *maxx(int *L, int n);
int *path(int *L, int *OPT, int n);

void print_array(int *L, int n) {
    printf("[");
    for (int i = 0; i < n; i++) {
        printf("%d, ", L[i]);
    }
    printf("%d]\n", L[n]);
}
int max(int a, int b) {
    if (a >= b) {
        return a;
    } else
        return b;
}

int *maxx(int *L, int n) {
    int *OPT;
    OPT = malloc(n * (sizeof(int)));
    OPT[0] = L[0];
    OPT[1] = max(L[0], L[1]);
    for (int j = 2; j < n; j++) {
        OPT[j] = max(OPT[j - 1], (OPT[j - 2] + L[j]));
    }
    return OPT;
}

int *path(int *L, int *OPT, int n) {
    int *S = malloc((n / 2) * (sizeof(int)));
    int j = n;
    int i = 0;
    while (j >= 3) {
        if (OPT[j - 1] >= (L[j] + OPT[j - 2])) {
            j--;
        } else {
            S[i] = L[j];
            i++;
            j -= 2;
        }
    }
    if ((j == 2) && L[0] < L[1]) {
        S[i] = L[1];
    } else {
        S[i] = L[0];
    }
    return S;
}

int main() {
    int list[] = {1, 4, 8, 4, 3, 10};
    int n = sizeof(list) / sizeof(int);
    int *opt = maxx(list, n);
    int *S = path(list, opt, n);
    int n_s = n / 2;

    print_array(S, n_s);
}

File: secondo_anno/Algoritmi/Modulo_due/programmazione_dinamica/Lezioni/Insieme_Indipendente/cammino.py
'''
def maxx(L):
    OPT = [0] * len(L)
    OPT[0] = L[0]
    OPT[1] = max(L[0], L[1])
    for j in range(2, len(L)):
        if OPT[j - 1] > (OPT[j - 2] + L[j]):
            OPT[j] = OPT[j - 1]
        else:
            OPT[j] = (OPT[j - 2] + L[j])
    return OPT[-1]
'''

'''
In max2() andiamo a calcolare la sequenza ottima e ritorniamo l'array OPT, mentre in path ritorniamo S,
che è una lista che contiene liste lunghe due dove in posziione 0 c'è l'indice dell'elemento nella lista L iniziale
e in posizione 1 il valore effettivo per arrivare all'ottimo.
'''
def max2(L):
    OPT = [0] * len(L)
    OPT[0], OPT[1] = L[0], max(L[0], L[1])
    for j in range(2, len(L)):
        OPT[j] = max(OPT[j - 1], (OPT[j - 2] + L[j]))
    return OPT
l = [1, 4, 8, 4, 3, 10]
print(max2(l))



def path(L, OPT):
    S = []
    j = len(L) - 1
    while j >= 2:
        if OPT[j - 1] >= (OPT[j - 2] + L[j]):
            j -=1
        else:
            S.append([j, L[j]])
            j -= 2
    if j == 2 and L[0] < L[1]:
        S.append([1, L[1]])
    else:
        S.append([0, L[0]])
    return S
l = [1, 4, 8, 4, 3, 10]
OPT = max2(l)
print(path(l, OPT))
